const chalk = require('chalk');
const { VISIT } = require('../frame');

async function Test(name, timeout, customErrorMessage, isCustomErrorMessageVisible, isErrorDescriptionVisible, isStackVisible, points, callback, url) {

  await VISIT(url);

  return new Promise((resolve) => {
    const timeoutId = setTimeout(() => {
      console.log(chalk.red('‚Äº A teszt timeoutja lej√°rt ‚Äº'));
      //! TMP
      // process.exit(1);
    }, timeout);

    callback().then(
      () => {
        clearTimeout(timeoutId);
        console.log(`‚úî ${chalk.green('Sikeresen lefutott a teszt!')}`);
        // TODO: valahol t√°rolni kellene a pontokat
        resolve(points);
      }).catch(
      (err) => {
        if (isStackVisible) {
          console.log(chalk.yellow('üì¢ Stack trace:'));
          console.log(err.stack);
        }

        if (isErrorDescriptionVisible) {
          console.log(chalk.red.bold(err.message));
        }

        if (err.name === 'TestFailedError') {
          if (isCustomErrorMessageVisible) {
            clearTimeout(timeoutId);
            console.log(`‚ùå ${chalk.red(customErrorMessage)}`);
          }
          resolve(0);
        }
      });
  });
}

module.exports = Test;